src/main/java/hudson/remoting/Launcher.java

```diff
@@ -326,95 +407,38 @@ public void run() throws Exception {
-                @Override public PasswordAuthentication getPasswordAuthentication() {
-                    return new PasswordAuthentication(auth.substring(0,idx), auth.substring(idx+1).toCharArray());
-                }
-            });
-        }
-        if (candidateCertificates != null && !candidateCertificates.isEmpty()) {
-            HttpsURLConnection.setDefaultSSLSocketFactory(getSSLSocketFactory());
+        createX509Certificates();
+        try {
+            sslSocketFactory = Engine.getSSLSocketFactory(x509Certificates);
+        } catch (GeneralSecurityException | PrivilegedActionException e) {
+            throw new RuntimeException(e);
         }
         if(connectionTarget!=null) {
             runAsTcpClient();
         } else
-        if(agentJnlpURL !=null) {
-            List<String> jnlpArgs = parseJnlpArguments();
-            if (jarCache != null) {
-              jnlpArgs.add("-jar-cache");
-              jnlpArgs.add(jarCache.getPath());
-            }
-            if (this.noReconnect) {
-                jnlpArgs.add("-noreconnect");
-            }
-            if (this.noKeepAlive) {
-                jnlpArgs.add("-noKeepAlive");
-            }
-            if (agentLog != null) {
-                jnlpArgs.add("-agentLog");
-                jnlpArgs.add(agentLog.getPath());
-            }
-            if (loggingConfigFilePath != null) {
-                jnlpArgs.add("-loggingConfig");
-                jnlpArgs.add(loggingConfigFilePath.getAbsolutePath());
-            }
-            if (this.workDir != null) {
-                jnlpArgs.add("-workDir");
-                jnlpArgs.add(workDir.getAbsolutePath());
-                jnlpArgs.add("-internalDir");
-                jnlpArgs.add(internalDir);
-                if (failIfWorkDirIsMissing) {
-                    jnlpArgs.add("-failIfWorkDirIsMissing");
-                }
-            }
-            if (candidateCertificates != null && !candidateCertificates.isEmpty()) {
-                for (String c: candidateCertificates) {
-                    jnlpArgs.add("-cert");
-                    jnlpArgs.add(c);
-                }
-            }
-            if (noCertificateCheck) {
-                // Generally it is not required since the default settings have been changed anyway.
-                // But we set it up just in case there are overrides somewhere in the logic
-                jnlpArgs.add("-disableHttpsCertValidation");
-            }
-            try {
-                hudson.remoting.jnlp.Main._main(jnlpArgs.toArray(new String[0]));
-            } catch (CmdLineException e) {
-                System.err.println("JNLP file "+ agentJnlpURL +" has invalid arguments: "+jnlpArgs);
-                System.err.println("Most likely a configuration error in the controller");
-                System.err.println(e.getMessage());
-                System.exit(1);
+        if (agentJnlpURL != null || !urls.isEmpty()) {
```

Comment:
I think this also needs to cover `directConnection`, where neither `-jnlpUrl` nor `-url` is used yet we are running an inbound agent. https://github.com/jenkinsci/kubernetes-plugin/pull/1451#issuecomment-1775147649

Reply:
Indeed, somehow I missed testing this mode during all the other manual testing I did. Thanks for fixing this so quickly.